// @secretlint/formatter/module/index.js
import { loadFormatter as textlintCreateFormatter, getFormatterList as textlintGetFormatterList, } from "@textlint/linter-formatter";
import terminalLink from "terminal-link";
import { moduleInterop } from "@textlint/module-interop";
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
// @ts-expect-error: no @types
import isFile from "is-file";
// @ts-expect-error: no @types
import tryResolve from "try-resolve";
import debug0 from "debug";
import url from "node:url";
const debug = debug0("@secretlint/formatter");
const __dirname = path.dirname(fileURLToPath(import.meta.url));
/**
 * Convert secretlint result to textlint result for formatter
 */
const convertSecretLintResultToTextlintResult = (secretLintCoreResult, { enableTerminalLink }) => {
    return {
        filePath: secretLintCoreResult.filePath,
        messages: secretLintCoreResult.messages.map((message) => {
            const severityLevel = message.severity === "info"
                ? 0
                : message.severity === "warning"
                    ? 1
                    : message.severity === "error"
                        ? 2
                        : 0;
            // If the message has docsUrl, try to link to docsUrl
            const messageId = enableTerminalLink && message.docsUrl
                ? terminalLink(message.messageId, message.docsUrl, {
                    fallback: (text, _url) => {
                        return text;
                    },
                })
                : message.messageId;
            return {
                // Preset rule format
                // {preset-id} > {rule-id}
                // Single rule format
                // {rule-id}
                ruleId: message.ruleParentId ? `${message.ruleParentId} > ${message.ruleId}` : message.ruleId,
                /**
                 * @deprecated backword compatibility - use range or loc
                 */
                index: message.range[0],
                /**
                 * @deprecated backword compatibility - use range or loc
                 */
                line: message.loc.start.line,
                /**
                 * @deprecated backward compatibility - use range or loc
                 */
                column: message.loc.start.column,
                range: message.range,
                loc: message.loc,
                severity: severityLevel,
                message: `[${messageId}] ${message.message}`,
                data: message.data,
                // NO NEED - DUMMY DATA
                fix: undefined,
                type: "lint",
            };
        }),
    };
};
export async function loadFormatter(formatterConfig) {
    const formatterName = formatterConfig.formatterName;
    const isHumanReadableFormat = ["stylish", "pretty-error"].includes(formatterName);
    /**
     * Terminal Link is enabled when use human-readable format and option is enabled
     */
    const enableTerminalLink = isHumanReadableFormat ?? formatterConfig.terminalLink ?? false;
    debug(`formatterName: ${formatterName}`);
    try {
        const formatter = await secretlintCreateFormatter(formatterConfig);
        return {
            format: (results) => {
                return formatter.format(results);
            },
        };
    }
    catch {
        const formatter = await textlintCreateFormatter(formatterConfig);
        return {
            format: (results) => {
                return formatter.format(results.map((result) => convertSecretLintResultToTextlintResult(result, {
                    enableTerminalLink,
                })));
            },
        };
    }
}
export async function secretlintCreateFormatter(formatterConfig) {
    const formatterName = formatterConfig.formatterName;
    debug(`formatterName: ${formatterName}`);
    let formatter;
    let formatterPath;
    if (fs.existsSync(formatterName)) {
        formatterPath = formatterName;
    }
    else if (fs.existsSync(path.resolve(process.cwd(), formatterName))) {
        formatterPath = path.resolve(process.cwd(), formatterName);
    }
    else {
        if (isFile(`${path.join(__dirname, "formatters", formatterName)}.js`)) {
            formatterPath = `${path.join(__dirname, "formatters", formatterName)}.js`;
        }
        else if (isFile(`${path.join(__dirname, "formatters", formatterName)}.ts`)) {
            formatterPath = `${path.join(__dirname, "formatters", formatterName)}.ts`;
        }
        else {
            const pkgPath = tryResolve(formatterName) || tryResolve(`secretlint-formatter-${formatterName}`);
            if (pkgPath) {
                formatterPath = pkgPath;
            }
        }
    }
    try {
        // dynamic import require file url
        const fileUrl = url.pathToFileURL(formatterPath).href;
        formatter = moduleInterop(await import(fileUrl)).default;
    }
    catch (ex) {
        throw new Error(`Could not find formatter ${formatterName}
${ex}`);
    }
    return {
        format: function secretlintFormatterFormat(results) {
            return formatter(results, formatterConfig);
        },
    };
}
export function getFormatterList() {
    const textlintFormatters = textlintGetFormatterList();
    const secretlintFormatters = secretlintGetFormatterList();
    const secretlintFormatterNames = secretlintFormatters.map((formatter) => formatter.name);
    const externalFormatters = textlintFormatters.filter((formatter) => !secretlintFormatterNames.includes(formatter.name));
    return [...externalFormatters, ...secretlintFormatters];
}
export function secretlintGetFormatterList() {
    return fs
        .readdirSync(path.join(__dirname, "formatters"))
        .filter((file) => {
        return [".js", ".ts"].some((formatterExtension) => path.extname(file) === formatterExtension && !file.endsWith(".d.ts"));
    })
        .map((file) => {
        const nameWithoutExtension = [".js", ".ts"].reduce((name, extension) => path.basename(name, extension), file);
        return { name: nameWithoutExtension };
    });
}
